File: auth/src/main/java/com/webasyst/auth/WebasystAuthService.java
package com.webasyst.auth;

import android.app.PendingIntent;
import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Consumer;

import net.openid.appauth.AuthState;
import net.openid.appauth.AuthorizationRequest;
import net.openid.appauth.AuthorizationService;
import net.openid.appauth.AuthorizationServiceConfiguration;
import net.openid.appauth.CodeVerifierUtil;
import net.openid.appauth.ResponseTypeValues;
import net.openid.appauth.TokenRequest;

public class WebasystAuthService {
    private static final String TAG = "WA_AUTH_SERVICE";

    private final WebasystAuthStateStore stateStore;
    private final WebasystAuthConfiguration configuration;
    private final AuthorizationServiceConfiguration authServiceConfiguration;
    private final AuthorizationService authorizationService;

    @Nullable
    static WebasystAuthConfiguration currentConfiguration = null;

    private WebasystAuthService(@NonNull final Context context,
                                @NonNull final WebasystAuthConfiguration configuration) {
        stateStore = WebasystAuthStateStore.getInstance(context);
        this.configuration = configuration;
        this.authServiceConfiguration = new AuthorizationServiceConfiguration(
            configuration.authEndpoint,
            configuration.tokenEndpoint
        );
        authorizationService = new AuthorizationService(context);
    }

    public static WebasystAuthService getInstance(@NonNull final Context context) {
        if (null == currentConfiguration) {
            throw new IllegalStateException("Configuration must be set");
        }
        return new WebasystAuthService(context, currentConfiguration);
    }

    private AuthorizationService getAuthorizationService() {
        return authorizationService;
    }

    /**
     * Configures WAID service. This method should be called once, before any interactions with WAID.
     *
     * The recommended point to do it is your Application's onCreate() method
     * (if you use custom Application class) or authentication activity's onCreate().
     *
     * @param configuration Configuration object
     */
    public static void configure(WebasystAuthConfiguration configuration) {
        currentConfiguration = configuration;
    }

    void performTokenRequest(TokenRequest request) {
        authorizationService.performTokenRequest(request, stateStore::updateAfterTokenResponse);
    }

    /**
     * Performs given task with fresh access token. Token is automatically refreshed if needed.
     */
    public <T> void withFreshAccessToken(final AccessTokenTask<T> task) {
        withFreshAccessToken(task, null);
    }

    /**
     * Performs given task with fresh access token. Token is automatically refreshed if needed.
     * This variant if {@link #withFreshAccessToken} calls the callback upon task completion.
     */
    public <T> void withFreshAccessToken(final AccessTokenTask<T> task, @Nullable final Consumer<T> callback) {
        Log.d(TAG, "Running task with fresh token...");
        stateStore.getCurrent().performActionWithFreshTokens(authorizationService,
            (accessToken, idToken, exception) -> {
                stateStore.writeCurrent();
                if (exception != null) {
                    Log.w(TAG, "Caught exception in withFreshToken()", exception);
                    if (exception.code >= 2000) {
                        stateStore.replace(new AuthState());
                    }
                }
                final T result = task.apply(accessToken, exception);
                if (null != callback) callback.accept(result);
            });
    }

    /**
     * Performs sign in
     *
     * @param request Sign in request
     * @param success {@link PendingIntent} to be called upon successful sign in
     * @param cancelled {@link PendingIntent} to be called upon sign in cancellation
     */
    public void signIn(AuthorizationRequest request, PendingIntent success, PendingIntent cancelled) {
        getAuthorizationService().performAuthorizationRequest(request, success, cancelled);
    }

    /**
     * Performs sign out
     */
    public void signOut() {
        stateStore.replace(new AuthState());
    }

    /**
     * Releases resources
     */
    public void dispose() {
        authorizationService.dispose();
    }

    AuthorizationRequest createAuthorizationRequest() {
        final String codeVerifier = CodeVerifierUtil.generateRandomCodeVerifier();

        return new AuthorizationRequest.Builder(
            authServiceConfiguration,
            configuration.clientId,
            ResponseTypeValues.CODE,
            configuration.callbackUri
        )
            .setCodeVerifier(codeVerifier)
            .setScopes(configuration.scope)
            .build();
    }

    /**
     * This interface represents task to be used with {@link #withFreshAccessToken}.
     *
     * @param <T>
     */
    public interface AccessTokenTask<T> {
        T apply(@Nullable String accessToken, @Nullable Exception exception);
    }
}

File: webasyst-x/src/main/java/auth/WebasystAuthService.kt
package com.webasyst.x.auth

import android.app.PendingIntent
import android.content.Context
import android.net.Uri
import com.webasyst.x.util.SingletonHolder
import net.openid.appauth.AuthState
import net.openid.appauth.AuthorizationRequest
import net.openid.appauth.AuthorizationService
import net.openid.appauth.AuthorizationServiceConfiguration
import net.openid.appauth.CodeVerifierUtil
import net.openid.appauth.ResponseTypeValues
import net.openid.appauth.TokenRequest
import kotlin.coroutines.suspendCoroutine

class WebasystAuthService internal constructor(
    context: Context,
    private val clientId: String,
    authEndpoint: String,
    tokenEndpoint: String,
    callbackUri: String,
) : WebasystAuthStateManager.AuthStateObserver {
    private val callbackUri = Uri.parse(callbackUri)
    private val authService = AuthorizationService(context)
    private val authServiceConfiguration = AuthorizationServiceConfiguration(
        Uri.parse(authEndpoint),
        Uri.parse(tokenEndpoint)
    )

    private val stateStore = WebasystAuthStateManager.getInstance(context)

    init {
        stateStore.addObserver(this)
    }

    override fun onChange(state: AuthState) {
        if (state.needsTokenRefresh && state.refreshToken != null) {
            performTokenRequest(state.createTokenRefreshRequest())
        }
    }

    fun authorize(authRequest: AuthorizationRequest, pi: PendingIntent) {
        authService.performAuthorizationRequest(authRequest, pi)
    }

    fun logoff() {
        stateStore.replace(AuthState())
    }

    fun createAuthorizationRequest(): AuthorizationRequest {
        val codeVerivier = CodeVerifierUtil.generateRandomCodeVerifier()

        return AuthorizationRequest.Builder(
            authServiceConfiguration,
            clientId,
            ResponseTypeValues.CODE,
            callbackUri
        )
            .setCodeVerifier(codeVerivier)
            .build()
    }

    fun performTokenRequest(request: TokenRequest) {
        authService.performTokenRequest(request) { response, exception ->
            stateStore.updateAfterTokenResponse(response, exception)
        }
    }

    suspend fun <T> withFreshToken(block: suspend (accessToken: String?) -> T): T {
        val token = suspendCoroutine<String> { continuation ->
            stateStore.getCurrent().performActionWithFreshTokens(authService) { accessToken, _, ex ->
                continuation.resumeWith(when {
                    accessToken != null -> Result.success(accessToken)
                    ex != null -> Result.failure(ex)
                    else -> Result.failure(
                        IllegalStateException("Either token or exception should be set")
                    )
                })
            }
        }
        return block(token)
    }

    companion object : SingletonHolder<WebasystAuthService, Context>(::createWebasystAuthService) {
        const val EXTRA_AUTH_SERVICE_DISCOVERY = "authServiceDiscovery"

        internal val configuration = WebasystAuthConfiguration()
        fun configure(block: WebasystAuthConfiguration.() -> Unit) {
            configuration.apply(block)
        }
    }
}

private fun createWebasystAuthService(context: Context): WebasystAuthService = WebasystAuthService(
    context = context,
    clientId = WebasystAuthService.configuration.clientId ?: throw IllegalStateException("Client id must be set"),
    authEndpoint = WebasystAuthService.configuration.authEndpoint ?: throw IllegalStateException("Auth endpoint must be set"),
    tokenEndpoint = WebasystAuthService.configuration.tokenEndpoint ?: throw IllegalStateException("Token endpoint must be set"),
    callbackUri = WebasystAuthService.configuration.callbackUri ?: throw IllegalStateException("Callback uri must be set")
)
